File: hotel_management/hotel_management/report/sales_invoice_report/sales_invoice_report.py
import frappe

def get_columns():
    return [
        {"label": "Invoice Number", "fieldname": "parent", "fieldtype": "Link", "options": "Sales Invoice", "width": 200},
        {"label": "Item Name", "fieldname": "item_name", "fieldtype": "Data", "width": 200},
        {"label": "Item Code", "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 200},
        {"label": "Quantity", "fieldname": "quantity", "fieldtype": "Float", "width": 100},
        {"label": "Rate", "fieldname": "rate", "fieldtype": "Currency", "width": 100},
        {"label": "UOM", "fieldname": "uom", "fieldtype": "Data", "width": 80},
        {"label": "Amount", "fieldname": "amount", "fieldtype": "Currency", "width": 120},
    ]

def get_data(filters=None):
    filters_dict = {"docstatus": 1}

    if filters and filters.get("item"):
        filters_dict["item_code"] = filters["item"]

    if filters and filters.get("date"):
        sales_invoices = frappe.get_all(
            "Sales Invoice",
            fields=["name"],
            filters={"posting_date": filters["date"], "docstatus": 1},
        )
        invoice_names = [si["name"] for si in sales_invoices]
        if invoice_names:
            filters_dict["parent"] = ["in", invoice_names]
        else:
            return []

    sales_invoice_items = frappe.get_all(
        "Sales Invoice Item",
        fields=["parent", "item_name", "item_code", "qty as quantity", "rate", "uom", "amount"],
        filters=filters_dict,
    )

    if not sales_invoice_items:
        return []

    if filters and filters.get("quantity_sort"):
        sales_invoice_items.sort(
            key=lambda x: x["quantity"], 
            reverse=(filters["quantity_sort"] == "Descending")
        )

    return sales_invoice_items

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

File: hotel_management/hotel_management/report/sales_invoice_report/sales_invoice_report.js
frappe.query_reports["Sales Invoice Report"] = {
    "filters": [
        {
            "fieldname": "item",
            "fieldtype": "Link",
            "label": __("Item"),
            "options": "Item"
        },
        {
            "fieldname": "date",
            "fieldtype": "Date",
            "label": __("Date")
        },
        {
            "fieldname": "quantity_sort",
            "fieldtype": "Select",
            "label": __("Quantity Sort"),
            "options": "Ascending\nDescending"
        }
    ],
   "formatter": function (value, row, column, data, default_formatter) {
    let formatted_value = default_formatter(value, row, column, data);
    if (!data) {
        return formatted_value;
    }
    if (data.amount > 9000) {
        formatted_value = `<span style="background-color:rgb(147, 249, 136);">${formatted_value}</span>`;
		//row.style.backgroundColor = "rgb(147, 249, 136)";
    }
    if (column.fieldname === "quantity" && data.quantity > 5) {
        formatted_value = `<span style="color: green;">${formatted_value}</span>`;
    }
    if (column.fieldname === "quantity" && data.quantity <= 5) {
        formatted_value = `<span style="color: orange;">${formatted_value}</span>`;
    }
    if (column.fieldname === "amount" && data.amount < 8000) {
        formatted_value = `<span style="color: red;">${formatted_value}</span>`;
    }

    return formatted_value;
}
};
